test/system/src/test/java/io/pravega/test/system/ControllerFailoverTest.java

```diff
@@ -99,27 +117,38 @@ public static void setup() throws InterruptedException, MarathonException, URISy
         log.debug("pravega host service details: {}", segUris);
     }
 
-    private static URI getTestControllerServiceURI() {
-        Service controllerService = new PravegaControllerService(TEST_CONTROLLER_SERVICE_NAME, null);
-        List<URI> ctlURIs = controllerService.getServiceDetails();
-        return ctlURIs.get(0);
-    }
 
-    private static URI getControllerURI() {
-        Service controllerService = new PravegaControllerService("controller", null);
-        List<URI> ctlURIs = controllerService.getServiceDetails();
-        return ctlURIs.get(0);
-    }
+    @Before
+    public void setup() {
+        Service zkService = Utils.createZookeeperService();
+        Assert.assertTrue(zkService.isRunning());
+        List<URI> zkUris = zkService.getServiceDetails();
+        log.info("zookeeper service details: {}", zkUris);
 
+        controllerService1 = Utils.createPravegaControllerService(zkUris.get(0));
+        if (!controllerService1.isRunning()) {
+            controllerService1.start(true);
+        }
 
-    private static void stopTestControllerService() {
-        log.info("Stopping test controller service");
-        Service controllerService = new PravegaControllerService(TEST_CONTROLLER_SERVICE_NAME, null);
-        controllerService.stop();
+        List<URI> conUris = controllerService1.getServiceDetails();
+        log.info("conuris {} {}", conUris.get(0), conUris.get(1));
+        log.debug("Pravega Controller service  details: {}", conUris);
+        // Fetch all the RPC endpoints and construct the client URIs.
+        List<String> uris;
+        if (Utils.isDockerLocalExecEnabled()) {
```

Comment:
There are lots of these environment specific blocks, please revisit all those and abstract away the behavior.

Reply:
answered above.

Reply:
Apart from in places where hdfs has to be deployed, and in Controllerfailovertest which accesses controller via `tcp://`  and `pravega://` ,  the abstraction is done in remaining places already using methods like `createZookeeperService ` .